### DISCUSSION TITLE
Proposal: Add a way to extend existing classes

### DISCUSSION THREAD

**Discussion by Reelix** [dotnet/csharplang#1852] [Language Ideas]
Say you have basic logging in your program, and in certain cases you want to write a console line as well. In this case, you have three viable options:

1.) Hook into the console, and log everything written to it
2.)
Console.WriteLine("Text Here");
Logger.Log("Text Here");

3.)
SomeFunc("Text Here");

void SomeFunc(string text)
{
Console.WriteLine(text);
Logger.Log(text);
}
- - - - - -
I propose a way to be able to extend the base classes, and in doing so allow the developer to add additional functionality without the need to add new classes, so you can eventually add a new class to your project, and go like:

Console.LogLine("Text Here");

This could possibly be done by creating a class of a new class type "extension" that requires an existing class type as a name, and in doing so allows you to add new methods onto those classes.

**Comment by HaloFour** [comment]
I believe this would fall under the discussion here: https://github.com/dotnet/csharplang/issues/164

**Comment by benaadams** [comment]
Extension Methods? https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods

**Comment by Reelix** [comment]
Existing extension methods can only be used on instances of existing classes, and not on a static type, so you can't just go  
  
Console c = new Console();
c.LogLine("Text here");

Since Console is a static class.

I'm also proposing a more refined version, so even if the above were possible, you wouldn't need to create a new instance of it as you would be extending the base class itself. Think of it as a neater version of downloading Console.cs, adding a new method, then using the modified class instead of the original.

**Comment by orthoxerox** [comment]
@Reelix so basically you want to add static methods to existing static classes?

For your specific use case I would actually recommend to use dependency injection instead. This way you would always have a single set of *instance* methods to invoke while being free to swap out their implementation.

Do yoy have another use case for your proposal?

**Comment by Austin-bryan** [comment]
C# 8 will have Extension Everything, where you can do extension static methods.

**Comment by HaloFour** [comment]
Extension Everything isn't slated for C# 8.0.  It's bundled along with a number of future ideas like concepts, shapes and type classes which the team wants to explore over the coming months.
